driver.go

```diff
@@ -17,19 +17,33 @@
 package mysql
 
 import (
+	"context"
 	"database/sql"
 	"database/sql/driver"
+	"errors"
 	"net"
 )
 
+var (
+	errInvalidUser   = errors.New("invalid Connection: User is not set or longer than 32 chars")
+	errInvalidAddr   = errors.New("invalid Connection: Addr config is missing")
+	errInvalidNet    = errors.New("invalid Connection: Only tcp is valid for Net")
+	errInvalidDBName = errors.New("invalid Connection: DBName config is missing")
+)
+
 // watcher interface is used for context support (From Go 1.8)
 type watcher interface {
 	startWatcher()
 }
 
 // MySQLDriver is exported to make the driver directly accessible.
 // In general the driver is used via the database/sql package.
-type MySQLDriver struct{}
+type MySQLDriver struct {
+}
+
+type MySQLConnector struct {
+	Cfg *Config
+}
```

Comment:
I prefer private type and constructor function.

```go
type mysqlConnector *Config

func NewConnector(cfg *Config) (driver.Connector, error) {
    if err := cfg.normalize(); err != nil {
        return nil, err
    }
    return mysqlConnector(cfg)
}
```

Then, we can move validation from `Connector.Connect()` to `cfg.normalize()`.
User can pass config object created by `ParseDSN()`.